tests/integration/test_sh_boot.py

```diff
@@ -196,3 +197,194 @@ def main():
+            path = os.path.join(path, "x" * padding)
+            try:
+                touch(path)
+            except (IOError, OSError) as e:
+                if e.errno == errno.ENAMETOOLONG:
+                    return True
+                raise e
+
+        return False
+
+    limit = find_max_length(seed_max=2 ** 16, is_too_long=file_path_too_long)
+    sys.stderr.write(">>> Calculated file path length limit of : {}\n".format(limit))
+    return limit
+
+
+@pytest.fixture(scope="module")
+def file_name_length_limit(
+    tmpdir_factory,  # type: Any
+    file_path_length_limit,  # type: int
+):
+    # type: (...) -> int
+
+    def file_name_too_long(length):
+        # type: (int) -> bool
+        path = str(tmpdir_factory.mktemp("td"))
+        try:
+            touch(os.path.join(path, "x" * length))
+        except (IOError, OSError) as e:
+            if e.errno == errno.ENAMETOOLONG:
+                return True
+            raise e
+
+        return False
+
+    limit = find_max_length(seed_max=file_path_length_limit, is_too_long=file_name_too_long)
+    sys.stderr.write(">>> Calculated file name length limit of: {}\n".format(limit))
+    return limit
+
+
+@pytest.fixture(scope="module")
+def shebang_length_limit(
+    tmpdir_factory,  # type: Any
+    file_path_length_limit,  # type: int
+):
+    # type: (...) -> int
+
+    def shebang_too_long(length):
+        # type: (int) -> bool
+        path = str(tmpdir_factory.mktemp("td"))
+        while len(path) < length - len("#!" + os.path.join("directory", "x")):
+            path = os.path.join(path, "directory")
+            try:
+                os.mkdir(path)
+            except OSError as e:
+                if e.errno != errno.EEXIST:
+                    raise e
+
+        sh_path = os.path.join(path, "x" * (length - len("#!" + path + os.sep)))
+        try:
+            os.unlink(sh_path)
+        except OSError as e:
+            if e.errno != errno.ENOENT:
+                raise e
+        os.symlink("/bin/sh", sh_path)
+
+        script = os.path.join(path, "script.sh")
+        with open(script, "w") as fp:
+            fp.write("#!{sh_path}\n".format(sh_path=sh_path))
+            fp.write("exit 0\n")
+        chmod_plus_x(script)
+        try:
+            return 0 != subprocess.call(args=[script])
+        except OSError as e:
+            if e.errno == errno.ENOEXEC:
+                return True
+            raise e
+
+    limit = find_max_length(
+        seed_max=file_path_length_limit - len(os.sep + "script.sh"), is_too_long=shebang_too_long
+    )
+    sys.stderr.write(">>> Calculated shebang length limit of: {}\n".format(limit))
+    return limit
+
+
+@pytest.mark.parametrize("execution_mode_args", EXECUTION_MODE_ARGS_PERMUTATIONS)
+def test_shebang_length_limit(
+    tmpdir,  # type: Any
+    execution_mode_args,  # type: List[str]
+    shebang_length_limit,  # type: int
+):
+    # type: (...) -> None
+
+    long_dir_name = "x" * shebang_length_limit
```

Comment:
Looking at the failing test is the shebang length limit computed the max allowed or the shortest that will fail?
```suggestion
    long_dir_name = "x" * shebang_length_limit - 1
```